### PR TITLE
168129992 support multiple tables in one sheet

### PR DESCRIPTION
**Pull request by gidjin**
## Description

These changes add support for sheets that contain two different kinds of data to parse. Corrects an issue that caused the first row of data for service areas to be missed. Adds a verification method and a minimal test for that. I didn't think it was worthwhile adding another xlsx for testing malformed spreadsheets further, though I could be convinced otherwise. These changes required pulling the csvWriter creation up into the main, which we want to do but it does not pull in the writing of the csv yet. That will need to be part of another PR.

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

```sh
go test ./cmd/parse_ratengine_data_ghc -v -count 1 
make bin/ghc-pricing-parser
ghc-pricing-parser --filename "Pricing Template (2019-09-13) - FAKE DATA.xlsx" --save
```

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/tree/master/docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/tree/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [#168129992](https://www.pivotaltracker.com/story/show/168129992) for this change

### CODE REVIEW

File: cmd/parse_ratengine_data_ghc/main.go [hunk 1]
```diff
@@ -138,24 +143,39 @@ func initDataSheetInfo() {
 	xlsxDataSheets[4] = xlsxDataSheetInfo{
 		description:    stringPointer("1b) Service Areas"),
 		outputFilename: stringPointer("1b_service_areas"),
-		process:        &parseServiceAreas,
-		verify:         &verifyServiceAreas,
+		processMethods: []xlsxProcessInfo{
+			{
+				process:    &parseDomesticServiceAreas,
+				adtlSuffix: stringPointer("domestic"),
```
Comment (reggieriser) [inline cmd/parse_ratengine_data_ghc/main.go]: FYI, there's a `swag` utility library (with a `swag.String` method) we've used in some other places for this type of thing -- mostly in the handlers.

File: cmd/parse_ratengine_data_ghc/main.go [hunk 2]
```diff
@@ -373,13 +395,20 @@ func process(params paramConfig, sheetIndex int, tableFromSliceCreator services.
...
-			return errors.Wrapf(err, " process error for sheet index: %d with description: %s", sheetIndex, description)
+	if len(xlsxInfo.processMethods) > 0 {
+		for _, p := range xlsxInfo.processMethods {
+			// Create CSV writer to save data to CSV file, returns nil if params.saveToFile=false
+			csvWriter := createCsvWriter(params.saveToFile, sheetIndex, params.runTime, p.adtlSuffix)
+			if csvWriter != nil {
+				defer csvWriter.close()
+			}
+			var callFunc processXlsxSheet
+			callFunc = *p.process
```
Comment (reggieriser) [inline cmd/parse_ratengine_data_ghc/main.go]: I think there's the potential of nil pointer dereference here if we're picking a specific sheet to process (don't see a check above in that case).  Probably safest just to do a nil check here.

File: cmd/parse_ratengine_data_ghc/parse_service_areas.go [hunk 3]
```diff
@@ -12,10 +12,10 @@ import (
 )
 
 // parseServiceAreas: parser for: 1b) Service Areas
-var parseServiceAreas processXlsxSheet = func(params paramConfig, sheetIndex int, tableFromSliceCreator services.TableFromSliceCreator) error {
+var parseDomesticServiceAreas processXlsxSheet = func(params paramConfig, sheetIndex int, tableFromSliceCreator services.TableFromSliceCreator, csvWriter *createCsvHelper) error {
 	// XLSX Sheet consts
-	const xlsxDataSheetNum int = 4          // 1b) Service Areas
-	const serviceAreaRowIndexStart int = 10 // start at row 10 to get the rates
+	const xlsxDataSheetNum int = 4         // 1b) Service Areas
+	const serviceAreaRowIndexStart int = 9 // start at row 10 to get the rates
```
Comment (reggieriser) [inline cmd/parse_ratengine_data_ghc/parse_service_areas.go]: Nit: Comment talks about "rates" but we're on service areas ... looks like copy pasta.

File: cmd/parse_ratengine_data_ghc/shared.go [hunk 4]
```diff
@@ -124,15 +124,20 @@ func removeWhiteSpace(stripString string) string {
 }
 
 // generateOutputFilename: generates filename using xlsxDataSheetInfo.outputFilename
-// with the folling fomat -- <id>_<outputFilename>_<time.Now().Format("20060102150405")>.csv
-func (x *xlsxDataSheetInfo) generateOutputFilename(index int, runTime time.Time) string {
+// with the follwing fomat -- <id>_<outputFilename>_<time.Now().Format("20060102150405")>.csv
```
Comment (reggieriser) [inline cmd/parse_ratengine_data_ghc/shared.go]: Nit: "follwing fomat" --> "following format"

File: cmd/parse_ratengine_data_ghc/main.go [hunk 5]
```diff
@@ -302,11 +322,13 @@ func main() {
 	err = db.Transaction(func(connection *pop.Connection) error {
 		if params.processAll == true {
 			for i, x := range xlsxDataSheets {
-				if x.process != nil {
-					dbErr := process(params, i, tableFromSliceCreator)
-					if dbErr != nil {
-						log.Printf("Error processing xlsxDataSheets %v\n", dbErr.Error())
-						return dbErr
+				for _, p := range x.processMethods {
```
Comment (reggieriser) [inline cmd/parse_ratengine_data_ghc/main.go]: I'm thinking this loop is causing us to go through the process method twice.  This calls `process` for each process method, but the `process` function itself appears to go through each process method itself already.

File: cmd/parse_ratengine_data_ghc/shared.go (hunk 4 above)
Comment (gidjin) [reply cmd/parse_ratengine_data_ghc/shared.go]: nice catch

File: cmd/parse_ratengine_data_ghc/main.go (hunk 5 above)
Comment (gidjin) [reply cmd/parse_ratengine_data_ghc/main.go]: Good catch, you are right this shouldn't be looping through here. I'll fix it

**Pull request merged**